include/fastscapelib/structured_grid.hpp

```diff
@@ -0,0 +1,232 @@
+            };
+
+            return priority[a] < priority[b];
+        }
+
+    }  // namespace detail
+
+
+    /**
+     * Status at grid boundary nodes.
+     */
+    class boundary_status
+    {
+    protected:
+
+        bool is_looped(node_status status);
+    };
+
+    inline bool boundary_status::is_looped(node_status status)
+    {
+        return status == node_status::looped_boundary;
+    }
+
+
+    //***************
+    //* Grid elements
+    //***************
+
+    /**
+     * Grid/mesh node.
+     */
+    struct node
+    {
+        std::size_t idx;     /**< Node index */
+        node_status status;  /**< Node status */
+    };
+
+    /**
+     * Neighbor node.
+     */
+    struct neighbor
+    {
+        std::size_t idx;     /**< Index of the neighbor node */
+        double distance;     /**< Distance to the neighbor node */
+        node_status status;  /**< Status at the neighbor node */
+
+        bool operator==(const neighbor& rhs) const
+        {
+            return (idx == rhs.idx) && (distance == rhs.distance) && (status == rhs.status);
+        }
+    };
+
+    namespace detail
+    {
+        /**
+         * Add a given offset (positive, negative or zero) to a grid node index.
+         *
+         * This utility function is mainly to avoid signed/unsigned conversion
+         * warnings. Use it carefully.
+         */
+        inline std::size_t add_offset(std::size_t idx, std::ptrdiff_t offset)
+        {
+            return static_cast<std::size_t>(static_cast<std::ptrdiff_t>(idx) + offset);
+        }
+    }
+
+    //***************************
+    //* Structured grid interface
+    //***************************
+
+    template <class G>
+    struct grid_inner_types;
+
+    /**
+     * @class structured_grid
+     * @brief Add a common interface to all structured grid types.
+     *
+     * This class only defines a basic interface for all structured grid types.
+     * It does not embed any data member, this responsibility
+     * is delegated to the inheriting classes.
+     *
+     * @tparam G Derived grid type.
+     */
+    template <class G>
+    class structured_grid
+    {
+    public:
+
+        using derived_grid_type = G;
+        using inner_types = grid_inner_types<derived_grid_type>;
+
+        using size_type = typename inner_types::size_type;
+        using shape_type = typename inner_types::shape_type;
+        using spacing_type = typename inner_types::spacing_type;
+        using neighbors_type = typename inner_types::neighbors_type;
+
+        using spacing_t = std::conditional_t<std::is_arithmetic<spacing_type>::value,
+                                            spacing_type,
+                                            const spacing_type&>;
+
+        using node_status_type = typename inner_types::node_status_type;
+
+        //shape_type shape() const noexcept;
+        size_type size() const noexcept;
+        spacing_t spacing() const noexcept;
```

Comment:
For convenience, it would be nice to also expose a `length_t length()`, which would return the total grid length along each axis, i.e., `spacing * (size - 1)`.

Grid constructors could also support providing length + size instead of spacing + size. This is mostly for more convenience as depending on the model setup one may prefer the former or the latter way to set the grid geometry. But we could add this later!

Reply:
agree!